client/src/app/process-details/process-details.component.ts

```diff
@@ -15,45 +15,176 @@ import {NgxNode} from './ngxNode';
+      console.log('deleteInputBiomaterialFromProcess', data);
+      this.refreshGraph();
+    });
+  }
+
+  removeProtocol(protocol: MetadataDocument) {
+    console.log('removing protocol', protocol);
+    const protocolId = this.getId(protocol);
+    this.ingestService.deleteProtocolFromProcess(this.processId, protocolId).subscribe(data => {
+      this.refreshGraph();
+    });
+  }
+
+  removeOutputBiomaterial(biomaterial: MetadataDocument) {
+    console.log('add output biomaterial', biomaterial);
+    const biomaterialId = this.getId(biomaterial);
+    this.ingestService.deleteOutputBiomaterialFromProcess(this.processId, biomaterialId).subscribe(data => {
+      console.log('deleteOutputBiomaterialFromProcess', data);
+      this.refreshGraph();
+    });
+  }
+
+  removeOutputFile(file: MetadataDocument) {
+    console.log('add output file', file);
+    const fileId = this.getId(file);
+    this.ingestService.deleteOutputFileFromProcess(this.processId, fileId).subscribe(data => {
+      console.log('deleteOutputFileFromProcess', data);
+      this.refreshGraph();
+    });
+  }
+
+  // TODO Check if POST with comma-delimited resource uri's payload will do linking to all resources in the uri list
+  addProtocols() {
+    const tasks = this.protocolsToAdd.map(protocol => this.ingestService.addProtocolToProcess(this.processId, this.getId(protocol)));
+    forkJoin(tasks).subscribe(
+      data => {
+        console.log('addProtocols', data);
+        this.protocolsToAdd = [];
+        this.refreshGraph();
+      }
+    );
+  }
+
+  addOutputFiles() {
+    const tasks = this.outputFilesToAdd.map(file => this.ingestService.addOutputFileToProcess(this.processId, this.getId(file)));
+    forkJoin(tasks).subscribe(
+      data => {
+        console.log('addOutputFiles', data);
+        this.outputFilesToAdd = [];
+        this.refreshGraph();
+      }
+    );
+  }
+
+  addInputBiomaterials() {
+    const tasks = this.inputBiomaterialsToAdd.map(biomaterial => {
+      const biomaterialId = this.getId(biomaterial);
+      return this.ingestService.addInputBiomaterialToProcess(this.processId, biomaterialId);
+    });
+
+    forkJoin(tasks).subscribe(
+      data => {
+        console.log('addInputBiomaterials', data);
+        this.inputBiomaterialsToAdd = [];
+        this.refreshGraph();
+      }
+    );
+  }
+
+  addOutputBiomaterials() {
+    const tasks = this.outputBiomaterialsToAdd.map(biomaterial => {
+      const biomaterialId = this.getId(biomaterial);
+      return this.ingestService.addOutputBiomaterialToProcess(this.processId, biomaterialId);
+    });
+
+    forkJoin(tasks).subscribe(
+      data => {
+        console.log('addOutputBiomaterialToProcess', data);
+        this.outputBiomaterialsToAdd = [];
+        this.refreshGraph();
+      }
+    );
+  }
+
+  // TODO Create endpoint in Core to return all protocols, inputBiomaterials, derivedBiomaterials,derivedFiles for a process
+  private refreshGraph() {
```

Comment:
Can we make `protocolsToAdd` etc. into an observable (maybe rename to something better) and make the graph observe them? Actually, all of the `add*` and `remove*` functions could observe right?